src/frequenz/client/common/_component.py

```diff
@@ -0,0 +1,204 @@
+            return None
+
+        return InverterType(component_metadata.type)
+
+    return None
+
+
+class ComponentCategory(Enum):
+    """Possible types of microgrid component."""
+
+    NONE = components_pb.ComponentCategory.COMPONENT_CATEGORY_UNSPECIFIED
+    """Unspecified component category."""
+
+    GRID = components_pb.ComponentCategory.COMPONENT_CATEGORY_GRID
+    """Grid component."""
+
+    METER = components_pb.ComponentCategory.COMPONENT_CATEGORY_METER
+    """Meter component."""
+
+    INVERTER = components_pb.ComponentCategory.COMPONENT_CATEGORY_INVERTER
+    """Inverter component."""
+
+    BATTERY = components_pb.ComponentCategory.COMPONENT_CATEGORY_BATTERY
+    """Battery component."""
+
+    EV_CHARGER = components_pb.ComponentCategory.COMPONENT_CATEGORY_EV_CHARGER
+    """EV charger component."""
+
+    CHP = components_pb.ComponentCategory.COMPONENT_CATEGORY_CHP
+    """CHP component."""
+
+
+def _component_category_from_protobuf(
+    component_category: components_pb.ComponentCategory.ValueType,
+) -> ComponentCategory:
+    """Convert a protobuf ComponentCategory message to ComponentCategory enum.
+
+    For internal-only use by the `microgrid` package.
+
+    Args:
+        component_category: protobuf enum to convert
+
+    Returns:
+        Enum value corresponding to the protobuf message.
+
+    Raises:
+        ValueError: if `component_category` is a sensor (this is not considered
+            a valid component category as it does not form part of the
+            microgrid itself)
+    """
+    if component_category == components_pb.ComponentCategory.COMPONENT_CATEGORY_SENSOR:
+        raise ValueError("Cannot create a component from a sensor!")
+
+    if not any(t.value == component_category for t in ComponentCategory):
+        return ComponentCategory.NONE
+
+    return ComponentCategory(component_category)
+
+
+@dataclass(frozen=True)
+class Component:
+    """Metadata for a single microgrid component."""
+
+    component_id: int
+    """The ID of this component."""
+
+    category: ComponentCategory
+    """The category of this component."""
+
+    type: ComponentType | None = None
+    """The type of this component."""
+
+    def is_valid(self) -> bool:
+        """Check if this instance contains valid data.
+
+        Returns:
+            `True` if `id > 0` and `type` is a valid `ComponentCategory`, or if `id
+                == 0` and `type` is `GRID`, `False` otherwise
+        """
+        return (
+            self.component_id > 0 and any(t == self.category for t in ComponentCategory)
+        ) or (self.component_id == 0 and self.category == ComponentCategory.GRID)
+
+    def __hash__(self) -> int:
+        """Compute a hash of this instance, obtained by hashing the `component_id` field.
+
+        Returns:
+            Hash of this instance.
+        """
+        return hash(self.component_id)
+
+
+class ComponentMetricId(Enum):
```

Comment:
Straight from the SDK...

Reply:
OK, again, if we want to first commit the code copied from the SDK as is, then great, we can merge as is, but I would make it very clear in the commit message and PR description that this is the intention.

Reply:
Nah, fuck that, I'll do it properly from the beginning

Reply:
Though, import paths like `from frequenz.client.common.microgrid.components.components import ComponentCategory` still make me doubt that recreating that structure is the right way

Reply:
No, I think `components/components.proto` should be translated to `components/__init__.py` in here, so it should be `frequenz.client.common.microgrid.components import ComponentCategory`.

Reply:
And how would the other proto files then be translated?

```
components/
  battery.proto
  components.proto
  ev_charger.proto
  fuse.proto
  grid.proto
  inverter.proto
  transformer.proto
```

Reply:
`components/battery.py`, etc. I think we just treat `xxx/xxx.proto` specially and have it as `xxx/__init__.py` instead.

Reply:
As far as I see this pattern is used in other dirs too, for example `microgrid/microgrid.proto`, and usually the `xxx/xxx.proto` file uses other `xxx/yyy.proto` files and not the other way around.

@tiyash-basu-frequenz FYI, just to confirm that his pattern is used consciously and not just a coincidence. It would be important to have a consistent scheme to make the python modules make sense and make sure we don't run into cyclic dependencies.

In the case of `microgrid/sensor/sensor.proto`, where there is only that file in the `microgrid/sensor/` directory, I would just add that file as a plain `microgrid/sensor.py`. If more files are added in the future we can rename `sensor.py` to sensor/__init__.py` and it will be transparent to the end user.

Reply:
This is a pattern we consciously use, and this is a result of how we want to define packages in our protobuf definitions. `microgrid/sensor/sensor.proto` means that the entities in `sensor.proto` are a member of the package `...microgrid.sensor`. The same applies to all other protobuf files.

For references, check these links: https://github.com/frequenz-floss/frequenz-api-common/pull/127#discussion_r1385230614, https://github.com/frequenz-floss/frequenz-api-common/pull/131